velox/exec/tests/HashJoinTest.cpp

```diff
@@ -69,163 +95,327 @@ class HashJoinTest : public HiveConnectorTestBase {
-      opts.vectorSize = leftBatchSize;
-      fuzzer_.setOptions(opts);
-      for (int i = 0; i < numLeftBatches; ++i) {
-        leftBatches.push_back(fuzzer_.fuzzRow(leftType));
+  // Make batches with random data.
+  std::vector<RowVectorPtr> makeBatches(
+      int32_t batchSize,
+      int32_t numBatches,
+      RowTypePtr rowType,
+      double nullRatio = 0.1) {
+    std::vector<RowVectorPtr> batches;
+    batches.reserve(numBatches);
+    if (batchSize != 0) {
+      fuzzerOpts_.vectorSize = batchSize;
+      fuzzerOpts_.nullRatio = nullRatio;
+      VectorFuzzer fuzzer = getFuzzer();
+      for (int32_t i = 0; i < numBatches; ++i) {
+        batches.push_back(fuzzer.fuzzRow(rowType));
       }
     } else {
-      for (int i = 0; i < numLeftBatches; ++i) {
-        leftBatches.push_back(RowVector::createEmpty(leftType, pool_.get()));
+      batches.push_back(RowVector::createEmpty(rowType, pool_.get()));
+    }
+    // NOTE: we generate a number of vectors with a fresh new fuzzer init with
+    // the same fix seed. The purpose is to ensure we have sufficient match if
+    // we use the row type for both build and probe inputs. Here we shuffle the
+    // built vectors to introduce some randomness during the join execution.
+    std::default_random_engine rng(1234);
+    std::shuffle(std::begin(batches), std::end(batches), rng);
+    return batches;
+  }
+
+  std::vector<RowVectorPtr> makeBatches(
+      int32_t numBatches,
+      std::function<RowVectorPtr(int32_t)> makeVector) {
+    std::vector<RowVectorPtr> batches;
+    batches.reserve(numBatches);
+    for (int32_t i = 0; i < numBatches; ++i) {
+      batches.push_back(makeVector(i));
+    }
+    return batches;
+  }
+
+  std::vector<RowVectorPtr> mergeBatches(
+      std::vector<RowVectorPtr>&& lhs,
+      std::vector<RowVectorPtr>&& rhs,
+      bool shuffle = false) {
+    std::vector<RowVectorPtr> mergedBatches;
+    mergedBatches.reserve(lhs.size() + rhs.size());
+    std::move(lhs.begin(), lhs.end(), std::back_inserter(mergedBatches));
+    std::move(rhs.begin(), rhs.end(), std::back_inserter(mergedBatches));
+    if (shuffle) {
+      std::default_random_engine rng(1234);
+      std::shuffle(std::begin(mergedBatches), std::end(mergedBatches), rng);
+    }
+    return mergedBatches;
+  }
+
+  // Make splits with each plan node having a number of source files.
+  SplitInput makeSpiltInput(
+      std::vector<core::PlanNodeId> nodeIds,
+      const std::vector<std::vector<std::shared_ptr<TempFilePath>>>& files) {
+    VELOX_CHECK_EQ(nodeIds.size(), files.size());
+    SplitInput splitInput;
+    for (int i = 0; i < nodeIds.size(); ++i) {
+      std::vector<exec::Split> splits;
+      splits.reserve(files[i].size());
+      for (const auto& file : files[i]) {
+        splits.push_back(exec::Split(makeHiveConnectorSplit(file->path)));
       }
+
+      splitInput.emplace(nodeIds[i], std::move(splits));
     }
+    return splitInput;
+  }
+
+  void runJoinTest(
```

Comment:
This method has a lot of parameters making it difficult to understand what it does and how to call it. Would you consider breaking up the parameter list int a small number of structs or using a builder pattern?

Reply:
Yes, that's the plan per our discussion.